housekeeping/src/main/java/com/hotels/housekeeping/service/impl/FileSystemHousekeepingService.java

```diff
@@ -48,87 +56,126 @@
-                  + "Ok to ignore. {}", cleanUpPath, e.getMessage());
-        } catch (Exception e) {
-          LOG.warn("Path '{}' was not deleted from the housekeeping database. {}", cleanUpPath, e.getMessage());
-        }
+    if (!cleanUpPathExists) {
+      // BEWARE the eventual consistency of your blobstore!
+      try {
+        LOG.info("Deleting path '{}' from housekeeping database", cleanUpPath);
+        legacyReplicaPathRepository.delete(cleanUpPath);
+      } catch (ObjectOptimisticLockingFailureException e) {
+        LOG
+            .debug("Failed to delete path '{}': probably already cleaned up by process running at same time. "
+                + "Ok to ignore. {}", cleanUpPath, e.getMessage());
+      } catch (Exception e) {
+        LOG.warn("Path '{}' was not deleted from the housekeeping database. {}", cleanUpPath, e.getMessage());
       }
-    } catch (IOException e) {
-      LOG.warn("Path '{}' was not deleted from the housekeeping database. {}", cleanUpPath, e.getMessage());
     }
   }
 
   @Override
   public void cleanUp(Instant referenceTime) {
+    ExecutorService executor = Executors.newFixedThreadPool(numberOfCleanupThreads);
     try {
       Pageable pageRequest = new PageRequest(0, fetchLegacyReplicaPathPageSize);
       Page<LegacyReplicaPath> page = legacyReplicaPathRepository
           .findByCreationTimestampLessThanEqual(referenceTime.getMillis(), pageRequest);
-      processPage(page);
-      while (page.hasNext()) {
-        pageRequest = pageRequest.next();
+      processPage(page, executor);
+      int pagesProcessed = 1;
+      int totalPages = page.getTotalPages();
+      // We keep fetching the first page while we have pages. Because we are deleting entries the number of pages
+      // changes so we use the total number of pages from the first request to loop over all pages once.
+      while (page.hasNext() && pagesProcessed < totalPages) {
         page = legacyReplicaPathRepository.findByCreationTimestampLessThanEqual(referenceTime.getMillis(), pageRequest);
-        processPage(page);
+        processPage(page, executor);
+        pagesProcessed++;
       }
-    } catch (Exception e) {
+    } catch (Throwable e) {
       throw new HousekeepingException(format("Unable to execute housekeeping at instant %d", referenceTime.getMillis()),
           e);
+    } finally {
+      executor.shutdownNow();
     }
   }
 
-  private void processPage(Page<LegacyReplicaPath> page) {
-    for (LegacyReplicaPath cleanUpPath : page) {
-      cleanUpPath = fixIncompleteRecord(cleanUpPath);
-      housekeepPath(cleanUpPath);
+  private void processPage(Page<LegacyReplicaPath> page, ExecutorService executor) throws Throwable {
+    List<Future<Void>> futures = new ArrayList<>(page.getSize());
+    for (final LegacyReplicaPath cleanUpPath : page) {
+      Future<Void> future = executor.submit(new Callable<Void>() {
```

Comment:
Could this be a `Runnable` instead of a `Callable`?

Reply:
I don't really like Runnable because it give me an untyped Future: `Future<?>`
Then again I also don't like returning null in the Callable, but for some uncertain reason I find that less of a problem.
Why would you prefer a Runnable?

Reply:
I dont really mind either way, `Runnable` just seemed more appropriate because nothing is returned but happy to stick with `Callable`

Reply:
yeah both not ideal :(